File: backend/app/services/offenders.py
from __future__ import annotations

from collections import defaultdict

from sqlalchemy.orm import Session

from app.schemas import GraphNode, OffenderIncident, OffenderMatch, OffenderProfile
from app.services.graph import get_graph
from app.services import snapshot as snap


def get_offender_profiles(db: Session) -> list[OffenderProfile]:
    # Prefer non-empty cache; an empty list is a stale snapshot (known AppSail issue).
    cached = snap.mem_get("offenders")
    if cached is None:
        cached = snap.db_get(db, "offenders")
    if cached is not None and len(cached) > 0:
        snap.mem_set("offenders", cached)
        return [OffenderProfile.model_validate(o) for o in cached]

    data = get_graph(db)
    by_id = {n.id: n for n in data.nodes}

    linked: dict[str, list[OffenderIncident]] = defaultdict(list)
    for e in data.edges:
        if e.kind != "ACCUSED_IN":
            continue
        src, tgt = by_id.get(e.source), by_id.get(e.target)
        if not src or not tgt:
            continue
        person = src if src.kind == "Person" else tgt if tgt.kind == "Person" else None
        incident = src if src.kind == "Incident" else tgt if tgt.kind == "Incident" else None
        if not person or not incident or not incident.meta:
            continue
        m = incident.meta
        linked[person.id].append(
            OffenderIncident(
                id=incident.id,
                docket=incident.label,
                district=incident.district,
                at=int(m.get("At", 0)),
                entry=str(m.get("Entry", "—")),
                target=str(m.get("Target", "—")),
                window=str(m.get("Window", "—")),
            )
        )

    profiles: list[OffenderProfile] = []
    sig_index: dict[str, list[tuple[GraphNode, str]]] = defaultdict(list)

    for pid, incidents in linked.items():
        if len(incidents) < 2:
            continue
        person = by_id[pid]
        incidents.sort(key=lambda i: i.at)
        counts: dict[str, int] = defaultdict(int)
        for i in incidents:
            counts[f"{i.entry} → {i.target}"] += 1
        signature = max(counts, key=counts.get)  # type: ignore[arg-type]
        districts = sorted({i.district for i in incidents})
        span = max(0, (incidents[-1].at - incidents[0].at) // 86_400_000)
        priors = int(person.meta.get("priors", len(incidents) - 1)) if person.meta else len(incidents) - 1
        sig_index[signature].append((person, districts[0] if districts else person.district))
        profiles.append(
            OffenderProfile(
                person=person,
                incidents=incidents,
                districts=districts,
                signature=signature,
                spanDays=span,
                priors=priors,
                matches=[],
            )
        )

    for p in profiles:
        matches = []
        for other, dist in sig_index.get(p.signature, []):
            if other.id == p.person.id:
                continue
            if dist in p.districts:
                continue
            matches.append(OffenderMatch(person=other, district=dist, shared=1))
        p.matches = matches[:5]

    profiles.sort(key=lambda p: len(p.incidents), reverse=True)
    trimmed = profiles[:80]
    # Persist so subsequent requests (and Person Intel) see profiles
    try:
        snap.publish(db, "offenders", [p.model_dump() for p in trimmed])
    except Exception:
        snap.mem_set("offenders", [p.model_dump() for p in trimmed])
    return trimmed

File: src/data/offenders.ts
import { getNetwork } from './network'
import type { GraphNode } from './types'

/**
 * Repeat-offender profiles (§7.2, "repeat-offender tracking").
 *
 * §7.2 asks for a person's incident edges "aggregated into a timeline/profile
 * view, with MO tags compared for similarity across jurisdictions", and the
 * challenge brief names repeat-offender tracking as a capability in its own
 * right. Both parts matter and neither works alone: a list of a suspect's cases
 * is a record search, and an MO signature with no person attached is a
 * statistic. The finding is the intersection — this person, this method, these
 * districts, this stretch of time.
 */

export interface OffenderIncident {
  id: string
  docket: string
  district: string
  at: number
  entry: string
  target: string
  window: string
}

export interface OffenderProfile {
  person: GraphNode
  incidents: OffenderIncident[]
  /** Districts the person's incidents span. */
  districts: string[]
  /** Most frequent entry→target pair — the behavioural signature. */
  signature: string
  /** Days between first and most recent linked incident. */
  spanDays: number
  priors: number
  /** Other people whose dominant signature matches, in a different district. */
  matches: { person: GraphNode; district: string; shared: number }[]
}

function readIncident(n: GraphNode): OffenderIncident | null {
  const m = n.meta
  if (!m) return null
  return {
    id: n.id,
    docket: n.label,
    district: n.district,
    at: Number(m.At ?? 0),
    entry: String(m.Entry ?? '—'),
    target: String(m.Target ?? '—'),
    window: String(m.Window ?? '—'),
  }
}

const signatureOf = (i: OffenderIncident) => `${i.entry} → ${i.target}`

let cache: OffenderProfile[] | null = null

/**
 * Everyone linked to more than one incident, ranked by how many.
 *
 * One offence is not a pattern, so a single-incident person is not a repeat
 * offender and does not belong in this list at all.
 */
export function getOffenderProfiles(): OffenderProfile[] {
  if (cache) return cache

  const { nodes, edges } = getNetwork()
  const byId = new Map(nodes.map((n) => [n.id, n]))

  const linked = new Map<string, OffenderIncident[]>()
  for (const e of edges) {
    if (e.kind !== 'ACCUSED_IN') continue
    const person = byId.get(e.source)?.kind === 'Person' ? byId.get(e.source) : byId.get(e.target)
    const incidentNode =
      byId.get(e.source)?.kind === 'Incident' ? byId.get(e.source) : byId.get(e.target)
    if (!person || !incidentNode || incidentNode.kind !== 'Incident') continue

    const incident = readIncident(incidentNode)
    if (!incident) continue
    if (!linked.has(person.id)) linked.set(person.id, [])
    linked.get(person.id)!.push(incident)
  }

  const draft = [...linked.entries()]
    .filter(([, list]) => list.length > 1)
    .map(([id, list]) => {
      const person = byId.get(id)!
      const incidents = [...list].sort((a, b) => b.at - a.at)

      const tally = new Map<string, number>()
      for (const i of incidents) tally.set(signatureOf(i), (tally.get(signatureOf(i)) ?? 0) + 1)
      const signature = [...tally.entries()].sort((a, b) => b[1] - a[1])[0]?.[0] ?? '—'

      const districts = [...new Set(incidents.map((i) => i.district))]
      const first = incidents[incidents.length - 1]?.at ?? 0
      const last = incidents[0]?.at ?? 0

      return {
        person,
        incidents,
        districts,
        signature,
        spanDays: Math.max(0, Math.round((last - first) / 864e5)),
        priors: Number(person.meta?.Priors ?? 0),
        matches: [] as OffenderProfile['matches'],
      }
    })

  /*
   * Signature matching across jurisdictions.
   *
   * A shared signature only says something when the two people are working in
   * DIFFERENT districts — within one district a common method is more likely to
   * be a local norm than a link between two offenders. That restriction is what
   * makes this §7.2's "compared for similarity across jurisdictions" rather
   * than a generic similarity join.
   */
  const bySignature = new Map<string, typeof draft>()
  for (const p of draft) {
    if (!bySignature.has(p.signature)) bySignature.set(p.signature, [])
    bySignature.get(p.signature)!.push(p)
  }

  for (const p of draft) {
    const peers = bySignature.get(p.signature) ?? []
    p.matches = peers
      .filter((q) => q.person.id !== p.person.id)
      .filter((q) => !q.districts.every((d) => p.districts.includes(d)))
      .map((q) => ({
        person: q.person,
        district: q.districts[0] ?? q.person.district,
        shared: q.incidents.filter((i) => signatureOf(i) === p.signature).length,
      }))
      .sort((a, b) => b.shared - a.shared)
      .slice(0, 5)
  }

  cache = draft.sort(
    (a, b) => b.incidents.length - a.incidents.length || b.districts.length - a.districts.length,
  )
  return cache
}

